### DISCUSSION TITLE
Package resolution based on Node.js version

### DISCUSSION THREAD

**Discussion by shalvah** [npm/feedback#445] [CLI] [answered]
Recently I wanted to upgrade mocha to v9, which drops support for Node.js 10.x. However, our project (a library) still has users on 10.x, so we still run CI on 10.x. There weren't any major API changes, so I changed the mocha constraint in package.json to `"^8.0.0 || ^9.0.0"` and tested locally (Node.js 15), and it worked on v9.

However, on pushing to CI,  I noticed moha v9 was getting installed, even on Node.js 10.x, despite the fact that mocha v9 sets its `engines` field to `>= 11`. I tried using `--engines-strict`, but npm just warned me of incompatible Node.js versions and exited.

Could npm rather try a lower version, if the highest available one fails an engine requirement? Is there already a way to achieve this? Composer (PHP) does this, and it makes it easy to support multiple versions of a package per environment. I imagine it would be a nontrivial change to npm's resolution logic, but surely it's doable?

**Comment by ljharb** [comment]
Have you tried with npm 7 latest? I'm relatively sure it takes engines into account already.

**Reply by shalvah** [reply]
Nope, just tried, and it didn't work.

`package.json`:

```json
{
  "dependencies": {
    "mocha": "^8.4.0 || ^9.0"
  }
}
```

Switched to Node.js 10.24.1, upgraded to npm 7

```
❯ npm i --engine-strict
npm ERR! code EBADENGINE
npm ERR! engine Unsupported engine
npm ERR! engine Not compatible with your version of node/npm: mocha@9.0.1
npm ERR! notsup Not compatible with your version of node/npm: mocha@9.0.1
npm ERR! notsup Required: {"node":">= 12.0.0"}
npm ERR! notsup Actual:   {"npm":"7.17.0","node":"v10.24.1"}
```

Without `--engine-strict`:

```
❯ npm i
npm WARN EBADENGINE Unsupported engine { package: 'mocha@9.0.1',
npm WARN EBADENGINE   required: { node: '>= 12.0.0' },
npm WARN EBADENGINE   current: { node: 'v10.24.1', npm: '7.17.0' } }

added 92 packages, and audited 93 packages in 1s

❯ npm list mocha
in@ C:\Users\shalvah\<projectpath>
`-- mocha@9.0.1
```

**Comment by MylesBorins** [comment, marked as answer]
@shalvah we will not likely change behavior on npm 6, but we could try and improve this for 7

Please open an issue over at https://github.com/npm/rfcs if you think this should be categorized as a new feature or an issue at https://github.com/npm/cli if you believe this qualifies more as a bug

/cc @npm/cli for visibility
